libmamba/include/mamba/core/problems_graph_creator.hpp

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2019, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_PROBLEMS_GRAPH_CREATOR_HPP
+#define MAMBA_PROBLEMS_GRAPH_CREATOR_HPP
+
+#include <string>
+#include <utility>
+#include <vector>
+#include <regex>
+#include <variant>
+#include <unordered_set>
+
+#include "mamba/core/pool.hpp"
+#include "mamba/core/problems_graph_util.hpp"
+#include "mamba/core/package_info.hpp"
+#include "mamba/core/util.hpp"
+#include "mamba/core/solver.hpp"
+
+extern "C"
+{
+#include "solv/solver.h"
+}
+
+namespace mamba
+{
+    class MProblemsGraphCreator
+    {
+    public:
+        using graph_t = MProblemsGraph<MNode, MEdge>;
+        using node_id = MProblemsGraph<MNode, MEdge>::node_id;
+        using solv_id_to_node_id = std::unordered_map<Id, node_id>;
+        using problems_list = std::vector<MSolverProblem>;
+        MProblemsGraphCreator(MPool* pool);
+
+        const graph_t& graph_from(const problems_list& problems);
```

Comment:
I would be in favor of having a constructor that takes an iterable of `MSolverProblem` and a function to resolve dependencies `SolvableId -> std::vector<SolvableId>` so that we can forget about `MPool`, `MSolverProblem`... once the object is created.

That way, we can also create such an object with custom data by plugin some artificial problems and a function to resolve dependencies.

Reply:
That sounds good!